R/functions.R

```diff
@@ -0,0 +1,46 @@
+
+#' @export
+#' @rdname properties
+lightness <- function (shades, values = NULL) UseMethod("lightness")
+
+#' @export
+#' @rdname properties
+lightness.function <- function(shades, values = NULL) {
+  orig_call <- match.call(expand.dots = FALSE)
+  orig_call[[1]] <- quote(lightness)
+  palette <- shades
+  palette_q <- substitute(shades)
+  orig_env <- parent.frame()
+  
+  wrapper <- function (...) {
+    inner_call <- match.call(expand.dots = FALSE)
+    inner_call[[1]] <- palette_q
+    result <- eval(inner_call, envir = orig_env)
+    orig_call$shades <- result
+    eval(orig_call)
+  }
+  
+  formals(wrapper) <- formals(palette)
+  class(wrapper) <- class(palette)
+   
+  return(wrapper)
+}
```

Comment:
This is some impressive call-wrangling, but as well as potentially introducing a lot of boilerplate for each property accessor function, it feels a bit fragile. Might a simpler wrapper like the one below work, and require fewer "tricks"?

```r
lightness <- function (shades, values = NULL)
{
    if (is.function(shades))
        function(...) .replaceProperty(shades(...), values, "Lab", 1)
    else
        .replaceProperty(shades(...), values, "Lab", 1)
}
```

Reply:
You're probably right. I think I was concerned to do something that could basically work for all the different functions at once, but I suspect your version is simpler... I would suggest leaving `lightness` et al. as they are, and then having `.replaceProperty` become a generic with methods for functions, ggplot scales etc. That will work for `saturation`, `lightness`, `brightness`, `chroma` and `hue`. `opacity` looks as if it will need its own subclass.